src/main/scala/is/hail/annotations/ReadRowsRDD.scala

```diff
@@ -1,23 +1,200 @@
+      off += 1
+      c
+    }
+  }
+
+  override def read(to: Array[Byte], toOff: Int, len: Int): Int = {
+    if (end == off)
+      -1
+    else {
+      val p = math.min(len, end - off)
+      System.arraycopy(a, off, to, toOff, p)
+      off += p
+      p
+    }
+  }
+
+  override def skip(n: Long): Long = {
+    assert(end != off)
+    assert(n <= Int.MaxValue)
+    val p = math.min(n.toInt, end - off)
+    off += p
+    p
+  }
+
+  override def available(): Int = end - off
+}
+
+class ArrayOutputStream(sizeHint: Int = 32) extends OutputStream {
+  var a: Array[Byte] = new Array[Byte](sizeHint)
+  var off: Int = 0
+
+  private def ensure(n: Int) {
+    val newEnd = off + n
+    if (newEnd > a.length) {
+      val newLength = math.max((a.length * 3) / 2, off + n)
+      val newA = new Array[Byte](newLength)
+      System.arraycopy(a, 0, newA, 0, off)
+      a = newA
+    }
+  }
+
+  def clear() {
+    off = 0
+  }
+
+  def write(b: Int) {
+    ensure(1)
+    a(off) = b.toByte
+    off += 1
+  }
+
+  override def write(from: Array[Byte], fromOff: Int, len: Int) {
+    ensure(len)
+    System.arraycopy(from, fromOff, a, off, len)
+    off += len
+  }
+}
+
+abstract class OutputBuffer {
+  def flush(): Unit
+
+  def writeByte(b: Byte): Unit
+
+  def writeFloat(f: Float): Unit
+
+  def writeDouble(d: Double): Unit
+
+  def writeBytes(mem: Long, off: Long, n: Int): Unit
+
+  def writeBoolean(b: Boolean) {
+    writeByte(b.toByte)
+  }
+
+  def writeInt(i: Int) {
+    var j = i
+    do {
+      var b = j & 0x7f
+      j >>>= 7
+      if (j != 0)
+        b |= 0x80
+      writeByte(b.toByte)
+    } while (j != 0)
   }
 
+  def writeLong(l: Long) {
+    var j = l
+    do {
+      var b = j & 0x7f
+      j >>>= 7
+      if (j != 0)
+        b |= 0x80
+      writeByte(b.toByte)
+    } while (j != 0)
+  }
+}
+
+object LZ4Buffer {
+  final val blockSize = 128 * 1024
+  // final val blockSize = 16
+}
+
+class LZ4OutputBuffer(out: OutputStream) extends OutputBuffer {
+  val buf: Array[Byte] = new Array[Byte](LZ4Buffer.blockSize)
+  var off: Int = 0
+
+  val comp = new Array[Byte](8 + LZ4Utils.maxCompressedLength(LZ4Buffer.blockSize))
+
+  private def writeBlock() {
+    if (off > 0) {
+      val compLen = LZ4Utils.compress(comp, 8, buf, off)
+      Memory.storeInt(comp, 0, compLen)
+      Memory.storeInt(comp, 4, off) // decompLen
+
+      out.write(comp, 0, 8 + compLen)
+
+      off = 0
+    }
+  }
+
+  def flush() {
+    writeBlock()
+  }
+
+  def writeByte(b: Byte) {
+    if (off + 1 > buf.length)
+      writeBlock()
+    Memory.storeByte(buf, off, b)
+    off += 1
+  }
+
+  def writeFloat(f: Float) {
+    if (off + 4 > buf.length)
+      writeBlock()
+    Memory.storeFloat(buf, off, f)
+    off += 4
+  }
+
+  def writeDouble(d: Double) {
+    if (off + 8 > buf.length)
+      writeBlock()
+    Memory.storeDouble(buf, off, d)
+    off += 8
+  }
+
+  def writeBytes(from0: Long, from0Off: Long, n0: Int) {
+    var from = from0 + from0Off
+    var n = n0
```

Comment:
`assert(n0 > 0)` ?

Reply:
I added `assert(n0 >= 0)`.  No reason you can't have e.g. a 0-length string.

Reply:
Sure